core/src/main/scala/kafka/log/remote/RemoteIndexCache.scala

```diff
@@ -37,88 +40,125 @@ object RemoteIndexCache {
  * @param logDir               log directory
  */
+@threadsafe
 class RemoteIndexCache(maxSize: Int = 1024, remoteStorageManager: RemoteStorageManager, logDir: String)
-  extends Logging with Closeable {
-
-  val cacheDir = new File(logDir, DirName)
-  @volatile var closed = false
-
-  val expiredIndexes = new LinkedBlockingQueue[Entry]()
-  val lock = new Object()
-
-  val entries: util.Map[Uuid, Entry] = new java.util.LinkedHashMap[Uuid, Entry](maxSize / 2,
-    0.75f, true) {
-    override def removeEldestEntry(eldest: util.Map.Entry[Uuid, Entry]): Boolean = {
-      if (this.size() > maxSize) {
-        val entry = eldest.getValue
-        // Mark the entries for cleanup, background thread will clean them later.
-        entry.markForCleanup()
-        expiredIndexes.add(entry)
-        true
-      } else {
-        false
-      }
-    }
-  }
+  extends Logging with AutoCloseable {
+  /**
+   * Directory where the index files will be stored on disk.
+   */
+  private val cacheDir = new File(logDir, DirName)
+  /**
+   * Represents if the cache is closed or not. Closing the cache is an irreversible operation.
+   */
+  private val closed: AtomicBoolean = new AtomicBoolean(false)
+  /**
+   * Unbounded queue containing the removed entries from the cache which are waiting to be garbage collected.
+   */
+  private val expiredIndexes = new LinkedBlockingQueue[Entry]()
+  /**
+   * Actual cache implementation that this file wraps around.
+   *
+   * The requirements for this internal cache is as follows:
+   * 1. Multiple threads should be able to read concurrently.
+   * 2. Fetch for missing keys should not block read for available keys.
+   * 3. Only one thread should fetch for a specific key.
+   * 4. Should support LRU policy.
+   *
+   * We use [[Caffeine]] cache instead of implementing a thread safe LRU cache on our own.
+   *
+   * Visible for testing.
+   */
+  private[remote] var internalCache: Cache[Uuid, Entry] = Caffeine.newBuilder()
+    .maximumSize(maxSize)
+    // removeListener is invoked when either the entry is invalidated (means manual removal by the caller) or
+    // evicted (means removal due to the policy)
+    .removalListener((_: Uuid, entry: Entry, _: RemovalCause) => {
+      // Mark the entries for cleanup and add them to the queue to be garbage collected later by the background thread.
+      entry.markForCleanup()
+      expiredIndexes.add(entry)
+    })
+    .build[Uuid, Entry]()
+
 
   private def init(): Unit = {
-    if (cacheDir.mkdir())
+    try {
+      Files.createDirectory(cacheDir.toPath)
       info(s"Created $cacheDir successfully")
+    } catch {
+      case _: FileAlreadyExistsException =>
+        info(s"RemoteIndexCache directory $cacheDir already exists. Re-using the same directory.")
+      case e: Exception =>
+        error(s"Unable to create directory $cacheDir for RemoteIndexCache.", e)
+        throw new IllegalArgumentException(e)
```

Comment:
Fair question. Since this exception is propagated all the way outside during server initialization, I was thinking of having a control of what exception we throw from this cache so that it can be handled. But I take my words back, and I think I will propagate it outside as `KafkaException`. Would that be ok?

Reply:
I just checked the server initialization code and seems like we just propagate the underlying exception outside without wrapping in KafkaException (probably because this exception doesn't go to clients). I will do the same.